library/std/src/f32.rs

```diff
@@ -293,6 +293,34 @@ impl f32 {
         unsafe { intrinsics::fmaf32(self, a, b) }
     }
 
+    /// Possibly-fused multiply-add. Computes `(self * a) + b` that can be
+    /// fused if the code generator determines that (a) the target instruction
+    /// set has support for a fused operation, and (b) that the fused operation
+    /// is more efficient than the equivalent, separate pair of mul and add
+    /// instructions.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// #![feature(float_mul_add_fast)]
+    /// let m = 10.0_f32;
+    /// let x = 4.0_f32;
+    /// let b = 60.0_f32;
+    ///
+    /// // 100.0
+    /// let abs_difference = (m.mul_add_fast(x, b) - ((m * x) + b)).abs();
+    ///
+    /// assert!(abs_difference <= f32::EPSILON);
+    /// ```
+    #[cfg(not(bootstrap))]
+    #[cfg_attr(not(bootstrap), rustc_allow_incoherent_impl)]
+    #[must_use = "method returns a new number and does not mutate the original value"]
+    #[inline]
+    #[unstable(feature = "float_mul_add_fast", issue = "none")]
+    pub fn mul_add_fast(self, a: f32, b: f32) -> f32 {
```

Comment:
I'd rather we not call so many things "fast". Standard fast-math has so much UB that it's nearly unusable, so at the request of some people in the community I split out separate intrinsics that offer optimizable semanatics but are safe: https://github.com/rust-lang/rust/pull/120718

Do you have examples of this intrinsic offering better optimizations than `fadd_algebraic` and `fmul_algebraic`?

Reply:
it could simply be `mul_add2`, tbh, given it is exactly multiply-add but allows 2 roundings.

Reply:
it is somewhat undesirable that we already used `mul_add` to indicate the 1-rounding variant because it is common to use `fma` as the fused (1-rounding) version and `mul_add` as the 1-or-2 rounding version.

Reply:
> because it is common to use fma as the fused (1-rounding) version and mul_add as the 1-or-2 rounding version.

Ah that explains the terrible names of the new intrinsics, I guess.^^ It is not self-explaining at all...

Reply:
@saethlin I think your `_algebraic` intrinsics are a better fit for type-based techniques like [fast-floats](https://github.com/bluss/fast-floats), but it interacts differently with its surrounding. For example, consider $a(b/a) + a$
```rs
#[inline(never)]
fn foo_mul_add_fast(a: f64, b: f64) -> f64 {
    a.mul_add_fast(b / a, a)
}

#[inline(never)]
fn foo_algebraic(a: f64, b: f64) -> f64 {
    use std::intrinsics::fadd_algebraic as fadd;
    use std::intrinsics::fmul_algebraic as fmul;
    fadd(fmul(a, b / a), a)
}
```
Then the first does what is I think is intuitively expected while the second reformulates to $(ab)/a + a$ without any fmadd.
```console
$ RUSTFLAGS='-Ctarget-cpu=skylake' cargo +stage1 asm foo_mul_add_fast
[...]
        vdivsd xmm1, xmm1, xmm0
        vfmadd231sd xmm0, xmm1, xmm0
$ RUSTFLAGS='-Ctarget-cpu=skylake' cargo +stage1 asm foo_algebraic
[...]
        vmulsd xmm1, xmm1, xmm0
        vdivsd xmm1, xmm1, xmm0
        vaddsd xmm0, xmm1, xmm0
```
These are meaningfully different in case of overflow/underflow. Consider calling `foo_*(x, x)`:
* if `x = 1e300_f64`, the former is exact while the latter is `inf`
* if `x = 1e-300_f64`, the former is exact while the latter is off by a factor of two.

Regarding name, I'm happy with whatever has consensus.

Reply:
That is interesting. Those intrinsics set the LLVM fast-math flags `reassoc` and `contract`, and as someone who never really studied floating-point details, it seems like those authorize creating the fused multiply-add. Is this a missing optimization in LLVM, or do I misunderstand?

Reply:
Yeah, this example produces identical code if we remove the division, so the contraction really does happen (but not at `opt-level=0` for `*_algebraic`, while it does use `vfmadd*` at `opt-level=0` when `llvm.fmuladd.*` is used). I don't know whether this transformation for `foo_algebraic` is justified for the microarchitecture (e.g., which execution unit will be tied up while OOO executing subsequent instructions) or not.

While initial lowering of `llvm.fmuladd` is keyed on `isFMAFasterThanFMulAndFAdd` (which depends on the target and FP mode), const-folding makes it dependent on optimization level. From [Limit FMul constant folding for fma simplifications](https://reviews.llvm.org/D67434):
> My understanding is that as long as it is a fmuladd, the code generator did not make a decision about fusing yet. IIUC there are 2 ways it can decide to fuse: 1) replace it with an fma intrinsic or lower it to a FMA in the backend.

That review spun off the [clarifying language](https://reviews.llvm.org/D67552) and [introduced the `@fmuladd_const_fmul` test](https://github.com/llvm/llvm-project/commit/f3ab99dcf8afa#diff-0f2df0f2960d65283247cddf3e7ffe319b36da72957df15c2eedbffdab01e6f9R503-R512) lowering `fmuladd` to `fadd` when the multiply part ($a, b$ in $a b + c$) is const.

Empirically, it is useful that`llvm.fmuladd` typically produces the same result in debug and optimized builds, but I can provoke the optimization dependence using const terms.
```rs
fn t_mul_add_fast(c: f64) -> f64 {
    let a = 0.1_f64;
    let b = 0.1_f64;
    a.mul_add_fast(b, c)
}
```
and calling as
```rs
    let x = std::hint::black_box(0.1_f64);
    let y = t_mul_add_fast(-x * x);
```
With `RUSTFLAGS='-Ctarget-cpu=skylake -Copt-level=0'`, this uses an FMA and produces `-8.326672684688674e-19`, but at `opt-level=1`, the const part is computed at compile-time, leaving only `vaddsd`, and thus evaluating to `0.0`. I think this removes any doubt about nondeterminism.

Reply:
I'll resolve this thread since I think it's now off-topic given that I've reduced scope to only adding the intrinsics. I think there are some design choices here if one wants to have `_algebraic` variants and use of `fence`; perhaps make a new issue or move discussion to Zulip?